### ISSUE TITLE
Infinite loop if no 'error' event is registered

### ISSUE THREAD

**Issue by piranna** [juliangruber/reconnect-core#19]
In an error occurs on underlying transport, on line https://github.com/juliangruber/reconnect-core/blob/db4556d7ae208537e6c23a4312d80e4bf0ef4f10/index.js#L44 it emits an 'error' event, but Node.js EventEmitter has the default behaviour that if an 'error' event is not registered it's thrown, so ```onDisconnect(err)``` on the next line is not being called. The error it's already being notified as the reason why it was disconnected so listening to the 'disconnect' event is just enought, so I think it's just enought to wrap the ```emitter.emit('error', err)``` on an empty try-catch and keep going.

**Comment by piranna** [comment]
Seems this could be related to issue #18.

**Comment by juliangruber** [comment]
how is this an infinite loop? and as always, a runnable gist demonstrating the problem is a big help

**Comment by piranna** [comment]
Here you have an screenshoot of the infinite loop:

![pantallazo-4](https://cloud.githubusercontent.com/assets/532414/7317762/8f942f02-ea84-11e4-9372-9a83db047932.png)

At the left you see the error events the same as at the top-right. Line 5556 is where the error is notified, that you can see in the call stack that it's dispatched by the highlighted line, that's the one of reconnect-core. The point is that since the 'error' event is not registered and due to this Node.js throw an exection, the next line onDisconnect() is never called, so the event 'disconnect' is not fired and the backoff is not started. To fix it it's just enough by wrapping the ```emitter.emit('error', err)``` by an empty try-catch, so if the user didn't registered an 'error' event, onDisconnect() can be called.

**Comment by juliangruber** [comment]
but the user _needs_ to register an error event, that's standard in node

**Comment by piranna** [comment]
I don't need to register to the error event if the default behaviour of reconnect-core is just to disconnect the underlying transport and dispatch a disconnect event (that has already the error, by the way) and this is just all that I need and I want. I could be able to register an empty noop function for the error event, but why to do so?

**Comment by juliangruber** [comment]
it's your obligation to show to reconnect that you're handling the error. if you're just ignoring it, that can be fine

**Comment by piranna** [comment]
I still think there's a bug on reconnect, because that makes the disconnect event not being dispatched, and this is more important since it already has the event itself. If you do so instead, you are dispatching the error twice.

**Comment by juliangruber** [comment]
ok, i'll reopen if you come with a gist that highlights that this is a problem

**Comment by piranna** [comment]
I've been revised thoroughtly this issue and definitely, the loop problem was an issue on my side, where some (old?) code was also managing the reconnection, so combined with the backoff mechanism of reconnect-core this lead to an infinite loop (when backoff finished it loop mechanish, I was creating a new instance). My honest apologizes for the inconvenience.

But I've found is that the onDisconect() method was being efectively being dispatched, but when the underlying transport (websocket-stream, in this case) is being closed, since I'm not listening to the 'error' event and If I would do so, it would then dispatched too.

![pantallazo-8](https://cloud.githubusercontent.com/assets/532414/7429537/eca4b9ea-f002-11e4-8ed2-15b5babfeeea.png)

On the other hand, since Node.js by default exits when there's an unhandled event, this makes that the reconnection mechanism doesn't work, so that's the reason why I still think that the 'error' event should not be mandatory and the .emit() call should be wrapped by a try-catch, since the error is already being notified on the 'disconnect' event. This way the behaviour can be the same both on Node.js and browser:

```Javascript
[piranna@Mabuk:~/workspace/kurento-client-js_bundleDependencies]
 (master) > npm test

> kurento-client@6.0.0-dev test /home/piranna/workspace/kurento-client-js_bundleDependencies
> scripts/test

Unknown file type: /home/piranna/workspace/kurento-client-js_bundleDependencies/test/index.html
Unknown file type: /home/piranna/workspace/kurento-client-js_bundleDependencies/test/readme.md


BasicPipeline
events.js:85
      throw er; // Unhandled 'error' event
            ^
Error: connect ECONNREFUSED
    at exports._errnoException (util.js:746:11)
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1000:19)
npm ERR! Test failed.  See above for more details.
```

Also, while looking at this I've found that at https://github.com/juliangruber/reconnect-core/blob/db4556d7ae208537e6c23a4312d80e4bf0ef4f10/index.js#L62 the error is not being passed to backoff so it could be able to notify to the user when it fails, so I think it should be given too.
